### ISSUE TITLE
Discrepency in code and paper related to HGRNBitAttention

### ISSUE THREAD

**Issue by loki-r** [ridgerchu/matmulfreellm#37]
The equations in the paper and the code don't match for the last equation.

The figure shows the last output equation as 
![image](https://github.com/user-attachments/assets/e4f1e990-afaa-4aab-9cec-dafdb82195a6)

But based on the code. It seems like this is the execution 
### $o_t^{'} = RMSNORM(g_t) * \sigma(h_t)$ 
instead of  
### $o_t^{'} = RMSNORM(h_t) * \sigma(g_t)$

Code path : 

- (g_norm is called with $g_t$ and $h_t$) : g_norm(self.g_proj(hidden_states), rearrange(o, 'b h l d -> b l (h d)'))

   > https://github.com/ridgerchu/matmulfreellm/blob/ec1c298ffa3db6436831f3e6d46f4e59d0b99194/mmfreelm/layers/hgrn_bit.py#L139 

- (X is $g_t$ and O is $h_t$) in FusedRMSNormSwishGate.forward(self, x, o, ...) 
   > https://github.com/ridgerchu/matmulfreellm/blob/ec1c298ffa3db6436831f3e6d46f4e59d0b99194/mmfreelm/layers/hgrn_bit.py#L139 

- (sigmoid is called on O which is $h_t$ instead of $g_t$) in _layer_norm_fwd_1pass_kernel  y = y * o * tl.sigmoid(o) 
   > https://github.com/ridgerchu/matmulfreellm/blob/ec1c298ffa3db6436831f3e6d46f4e59d0b99194/mmfreelm/modules/fused_norm_gate.py#L128

Seems like this is fixed in [HGRN - flash-linear-attention](https://github.com/sustcsonglin/flash-linear-attention/commit/a5baa970c917fcb558ad8efdd93381bd3903fba0) repository

Are the results with the inverted equation or with the fixed equation ?

**Comment by ridgerchu** [comment]
Hi, I think it is a bug, due to the HGRN api modifications. the sigma should be applied to g_t for better performance, but now it is applied to h_t. and our pre-trained model also still using sigma h_t... we will fix this problem in arxiv soon, and we believe that if applied to g_t it would be better performance compared with our current version.
